src/librbd/Operations.cc

```diff
@@ -759,36 +761,35 @@ int Operations<I>::snap_rollback(const cls::rbd::SnapshotNamespace& snap_namespa
   if (r < 0)
     return r;
 
-  RWLock::RLocker owner_locker(m_image_ctx.owner_lock);
+  C_SaferCond cond_ctx;
   {
-    // need to drop snap_lock before invalidating cache
-    RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
-    if (!m_image_ctx.snap_exists) {
-      return -ENOENT;
+    RWLock::RLocker owner_locker(m_image_ctx.owner_lock);
+    {
+      // need to drop snap_lock before invalidating cache
+      RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
+      if (!m_image_ctx.snap_exists) {
+        return -ENOENT;
+      }
+
+      if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+        return -EROFS;
+      }
+
+      uint64_t snap_id = m_image_ctx.get_snap_id(snap_namespace, snap_name);
+      if (snap_id == CEPH_NOSNAP) {
+        lderr(cct) << "No such snapshot found." << dendl;
+        return -ENOENT;
+      }
     }
 
-    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+    r = prepare_image_update(false);
+    if (r < 0) {
       return -EROFS;
     }
 
-    uint64_t snap_id = m_image_ctx.get_snap_id(snap_namespace, snap_name);
-    if (snap_id == CEPH_NOSNAP) {
-      lderr(cct) << "No such snapshot found." << dendl;
-      return -ENOENT;
-    }
+    execute_snap_rollback(snap_namespace, snap_name, prog_ctx, &cond_ctx);
   }
```

Comment:
@dillaman 
```
zhuzha:~/ceph/ceph.ci/build% PYTHONPATH=../src/test/pybind:../src/pybind:${PYTHONPATH} nosetests -v test_rbd:TestImage.test_rollback_with_resize
test_rbd.TestImage.test_rollback_with_resize ... /home/mgolub/ceph/ceph.ci/src/librbd/operation/ResizeRequest.cc: In function 'void librbd::operation::ResizeRequest<ImageCtxT>::send() [with ImageCtxT = librbd::ImageCtx]' thread 7fe17b7fe700 time 2017-11-14 18:32:11.241069
/home/mgolub/ceph/ceph.ci/src/librbd/operation/ResizeRequest.cc: 58: FAILED assert(image_ctx.owner_lock.is_locked())
```
I assume your change reveals the bug in `SnapshotRollbackRequest<I>::send_resize_image()`, which sends `ResizeRequest` without holding the owner lock, while `ResizeRequest::send()` expects this.

Reply:
Hmm -- not immediately sure how that's failing since ```Operation::snap_rollback``` should be holding the lock when it invokes ```Operation::execute_snap_rollback```

Reply:
I suppose the main difference that previously we were holding the owner lock until `SnapshotRollbackRequest` is complete (waiting while holding the lock) and although `ResizeRequest::send()` was checking for `owner_lock.is_locked()` in a different thread it still succeeded?

Reply:
Ah -- yeah, that's the issue. Thanks! I pushed a new commit to address it.